src/dawn/SIR/SIR.h

```diff
@@ -348,9 +361,47 @@ struct Value : NonCopyable {
 
   /// @brief Set the value
   template <class T>
-  void setValue(const T& value) {
-    valueImpl_ = make_unique<ValueImpl<T>>(value);
-    type_ = TypeInfo<T>::Type;
+  typename std::enable_if<!std::is_arithmetic<T>::value>::type
+  setValue(const T& value) {
+    if (type_ == TypeKind::None) {
+      valueImpl_ = make_unique<ValueImpl<T>>(value);
+      type_ = TypeInfo<T>::Type;
+    } else {
+      if (TypeInfo<T>::Type == type_) {
+        valueImpl_ = make_unique<ValueImpl<T>>(value);
+      } else {
+        //type mismatch
+        DAWN_ASSERT_MSG(false, "type mismatch trying to set sir::Value");
+      }
+    }
+  }
+
+  template <class T>
+  typename std::enable_if<std::is_arithmetic<T>::value>::type
```

Comment:
Why two overloads? Might we be able to set the value if the type is constructible from T?